runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/io/SdkByteReadChannelJVM.kt

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.io
+
+import kotlinx.coroutines.runBlocking
+import java.io.InputStream
+import java.util.*
+
+/**
+ * Create a blocking [InputStream] that blocks everytime the channel suspends at [SdkByteReadChannel.read]
+ */
+public fun SdkByteReadChannel.toInputStream(): InputStream = InputAdapter(this)
+
+private const val DEFAULT_READ_BYTES = 8192L
+private class InputAdapter(private val ch: SdkByteReadChannel) : InputStream() {
+
+    private val buffer = SdkBuffer()
+
+    override fun read(): Int {
+        if (ch.isClosedForRead && buffer.size == 0L) return -1
+
+        if (buffer.size == 0L) {
+            val rc = readBlocking()
+            if (rc == -1L) return -1
+        }
+
+        return buffer.readByte().toInt() and 0xff
+    }
+    override fun read(b: ByteArray, off: Int, len: Int): Int {
+        Objects.checkFromIndexSize(off, len, b.size)
+        if (ch.isClosedForRead && buffer.size == 0L) return -1
+        if (buffer.size == 0L) {
+            val rc = readBlocking()
+            if (rc == -1L) return -1
+        }
+
+        return buffer.read(b, off, len)
+    }
+
+    private fun readBlocking(): Long =
+        runBlocking {
+            ch.read(buffer, DEFAULT_READ_BYTES)
+        }
```

Comment:
**Question**: Looks like we _only_ ever read `DEFAULT_READ_BYTES` (8K) bytes at a time from the channel, even when the `read` call may have had a `len` higher than that. Given that `runBlocking` may be expensive, is it worth passing the requested length through to the channel read? Or making the chunk size configurable?

Reply:
Why do you feel that `runBlocking` is expensive?


------

The [len](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read-byte:A-int-int-) argument is documented as

> len - the maximum number of bytes to read.

We could change it but I chose 8K because thats the underlying segment size used by okio. I'm not sure this will matter all that much in practice but if you disagree I'm happy to change it.